src/psyclone/domain/gocean/kernel/psyir.py

```diff
@@ -0,0 +1,814 @@
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''This module contains PSyclone Kernel-layer-specific PSyIR classes.
+
+'''
+import re
+
+from fparser.common.readfortran import FortranStringReader
+from fparser.two import Fortran2003
+from fparser.two.utils import walk
+
+from psyclone.configuration import Config
+from psyclone.domain.gocean import GOceanConstants
+from psyclone.errors import InternalError
+from psyclone.parse.utils import ParseError
+from psyclone.psyir.symbols import DataTypeSymbol, Symbol, UnknownFortranType
+
+
+class KernelMetadataSymbol(DataTypeSymbol):
+    '''Specialise DataTypeSymbol to capture Kernel Metadata information,
+    verify that it conforms to the expected syntax and provide the
+    information to PSyclone in an easier to access form.
+
+    :param str name: the name of this symbol.
+    :param datatype: the type represented by this symbol.
+    :type datatype: :py:class:`psyclone.psyir.symbols.DataType`
+    :param visibility: the visibility of this symbol.
+    :type visibility: :py:class:`psyclone.psyir.symbols.Symbol.Visibility`
+    :param interface: the interface to this symbol.
+    :type interface: :py:class:`psyclone.psyir.symbols.SymbolInterface`
+
+    '''
+    def __init__(self, name, datatype, visibility=Symbol.DEFAULT_VISIBILITY,
+                 interface=None):
+        super().__init__(name, datatype, visibility, interface)
+        self.setup()
+
+    def setup(self):
+        '''Populate KernelMetadataSymbol properties and check that the symbol
+        contains valid metadata.
+
+        '''
+        if not isinstance(self.datatype, UnknownFortranType):
+            raise InternalError(
+                f"Expected kernel metadata to be stored in the PSyIR as "
+                f"an UnknownFortranType, but found "
+                f"{type(self.datatype).__name__}.")
+
+        # In an UnknownFortranType, the declaration is stored as a
+        # string, so create an fparser2 parse tree.
+        reader = FortranStringReader(self.datatype.declaration)
+        try:
+            spec_part = Fortran2003.Derived_Type_Def(reader)
+        except Fortran2003.NoMatchError:
+            # pylint: disable=raise-missing-from
+            raise InternalError(
+                f"Expected kernel metadata to be a derived type, but found "
+                f"'{self.datatype.declaration}'.")
+
+        const = GOceanConstants()
+        # Extract and store the required 'iterates_over',
+        # 'index_offset' and 'code' properties from the parse tree for
+        # ease of access.
+        value = self._get_property(spec_part, "iterates_over").string
+        self._validate_iterates_over(value)
```

Comment:
The setter also updates the underlying Fortran string representation to maintain consistency. At this point we are reading the underlying Fortran string representation and storing the relevant value. We could use the setter but it would then overwrite the string representation (with the same value) and I thought it better/clearer not to.